File: bukkit/src/main/java/io/wdsj/asw/bukkit/listener/packet/ASWBookPacketListener.java
package io.wdsj.asw.bukkit.listener.packet;

import com.github.retrooper.packetevents.event.PacketListenerAbstract;
import com.github.retrooper.packetevents.event.PacketListenerPriority;
import com.github.retrooper.packetevents.event.PacketReceiveEvent;
import com.github.retrooper.packetevents.protocol.packettype.PacketType;
import com.github.retrooper.packetevents.protocol.packettype.PacketTypeCommon;
import com.github.retrooper.packetevents.protocol.player.User;
import com.github.retrooper.packetevents.wrapper.play.client.WrapperPlayClientEditBook;
import io.wdsj.asw.bukkit.AdvancedSensitiveWords;
import io.wdsj.asw.bukkit.event.ASWFilterEvent;
import io.wdsj.asw.bukkit.event.EventType;
import io.wdsj.asw.bukkit.manage.notice.Notifier;
import io.wdsj.asw.bukkit.manage.permission.Permissions;
import io.wdsj.asw.bukkit.manage.punish.Punishment;
import io.wdsj.asw.bukkit.proxy.bungee.BungeeSender;
import io.wdsj.asw.bukkit.proxy.velocity.VelocitySender;
import io.wdsj.asw.bukkit.setting.PluginMessages;
import io.wdsj.asw.bukkit.setting.PluginSettings;
import io.wdsj.asw.bukkit.util.Utils;
import io.wdsj.asw.bukkit.util.cache.BookCache;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;

import static io.wdsj.asw.bukkit.AdvancedSensitiveWords.*;
import static io.wdsj.asw.bukkit.util.TimingUtils.addProcessStatistic;
import static io.wdsj.asw.bukkit.util.Utils.messagesFilteredNum;

public class ASWBookPacketListener extends PacketListenerAbstract {
    public ASWBookPacketListener() {
        super(PacketListenerPriority.LOW);
    }

    @Override
    public void onPacketReceive(PacketReceiveEvent event) {
        if (!isInitialized) return;
        if (!settingsManager.getProperty(PluginSettings.ENABLE_BOOK_EDIT_CHECK)) return;
        PacketTypeCommon packetType = event.getPacketType();
        User user = event.getUser();
        Player player = (Player) event.getPlayer();
        if (player.hasPermission(Permissions.BYPASS)) return;
        String userName = user.getName();
        if (packetType == PacketType.Play.Client.EDIT_BOOK) {
            String processedOutMessage = "";
            String outMessage = "";
            List<String> outList = new ArrayList<>();
            boolean skipReturnLine = settingsManager.getProperty(PluginSettings.BOOK_IGNORE_NEWLINE);
            boolean isCacheEnabled = settingsManager.getProperty(PluginSettings.BOOK_CACHE);
            boolean shouldSendMessage = false;
            boolean isCancelMode = settingsManager.getProperty(PluginSettings.BOOK_METHOD).equalsIgnoreCase("cancel");
            WrapperPlayClientEditBook wrapper = new WrapperPlayClientEditBook(event);

            // Book content check
            List<String> originalPages = wrapper.getPages();
            List<String> processedPages = new ArrayList<>(originalPages.size());
            long startTime = System.currentTimeMillis();
            for (String originalPage : originalPages) {
                if (skipReturnLine) originalPage = originalPage.replace("\n", "").replace("§0", "");
                if (settingsManager.getProperty(PluginSettings.PRE_PROCESS)) originalPage = originalPage.replaceAll(Utils.getPreProcessRegex(), "");
                boolean isBookCached = BookCache.isBookCached(originalPage);
                List<String> censoredWordList = isBookCached && isCacheEnabled ? BookCache.getCachedBookSensitiveWordList(originalPage) : AdvancedSensitiveWords.sensitiveWordBs.findAll(originalPage);
                String processedPage = isBookCached && isCacheEnabled ? BookCache.getCachedProcessedBookContent(originalPage) : AdvancedSensitiveWords.sensitiveWordBs.replace(originalPage);
                if (!censoredWordList.isEmpty()) {
                    if (!isBookCached && isCacheEnabled) BookCache.addToBookCache(originalPage, processedPage, censoredWordList);
                    if (isCancelMode) {
                        event.setCancelled(true);
                        shouldSendMessage = true;
                        outMessage = originalPage;
                        outList = censoredWordList;
                        processedOutMessage = processedPage;
                        break;
                    }

                    shouldSendMessage = true;
                    outMessage = originalPage;
                    outList = censoredWordList;
                    processedOutMessage = processedPage;
                }
                processedPages.add(processedPage);
            }
            if (!isCancelMode && shouldSendMessage) {
                wrapper.setPages(processedPages);
            }

            // Book title check
            String originalTitle = wrapper.getTitle();
            if (originalTitle != null) {
                if (settingsManager.getProperty(PluginSettings.PRE_PROCESS)) originalTitle = originalTitle.replaceAll(Utils.getPreProcessRegex(), "");
                List<String> censoredWordListTitle = AdvancedSensitiveWords.sensitiveWordBs.findAll(originalTitle);
                if (!censoredWordListTitle.isEmpty()) {
                    String processedTitle = AdvancedSensitiveWords.sensitiveWordBs.replace(originalTitle);
                    if (isCancelMode) {
                        event.setCancelled(true);
                    } else {
                        wrapper.setTitle(processedTitle);
                    }
                    shouldSendMessage = true;
                    outMessage = originalTitle;
                    outList = censoredWordListTitle;
                    processedOutMessage = processedTitle;
                }
            }

            if (shouldSendMessage) {
                messagesFilteredNum.getAndIncrement();
                if (settingsManager.getProperty(PluginSettings.ENABLE_API)) {
                    Bukkit.getPluginManager().callEvent(new ASWFilterEvent(player, outMessage, processedOutMessage, outList, EventType.BOOK, true));
                }
                if (settingsManager.getProperty(PluginSettings.HOOK_VELOCITY)) {
                    VelocitySender.send(player, EventType.BOOK, outMessage, outList);
                }
                if (settingsManager.getProperty(PluginSettings.HOOK_BUNGEECORD)) {
                    BungeeSender.send(player, EventType.BOOK, outMessage, outList);
                }
                if (settingsManager.getProperty(PluginSettings.ENABLE_DATABASE)) {
                    databaseManager.checkAndUpdatePlayer(player.getName());
                }
                long endTime = System.currentTimeMillis();
                addProcessStatistic(endTime, startTime);
                if (settingsManager.getProperty(PluginSettings.NOTICE_OPERATOR)) {
                    Notifier.notice(player, EventType.BOOK, outMessage, outList);
                }
                if (settingsManager.getProperty(PluginSettings.BOOK_PUNISH)) Punishment.punish(player);
                if (settingsManager.getProperty(PluginSettings.BOOK_SEND_MESSAGE)) {
                    player.sendMessage(ChatColor.translateAlternateColorCodes('&', messagesManager.getProperty(PluginMessages.MESSAGE_ON_BOOK)));
                }
                if (settingsManager.getProperty(PluginSettings.LOG_VIOLATION)) {
                    Utils.logViolation(userName + "(IP: " + user.getAddress().getAddress().getHostAddress() + ")(Book)",outMessage + outList);
                }
            }
        }
    }
}

File: bukkit/src/main/kotlin/io/wdsj/asw/bukkit/listener/packet/ASWBookPacketListener.kt
package io.wdsj.asw.bukkit.listener.packet

import com.github.retrooper.packetevents.event.PacketListenerAbstract
import com.github.retrooper.packetevents.event.PacketListenerPriority
import com.github.retrooper.packetevents.event.PacketReceiveEvent
import com.github.retrooper.packetevents.protocol.packettype.PacketType
import com.github.retrooper.packetevents.wrapper.play.client.WrapperPlayClientEditBook
import io.wdsj.asw.bukkit.AdvancedSensitiveWords
import io.wdsj.asw.bukkit.AdvancedSensitiveWords.settingsManager
import io.wdsj.asw.bukkit.manage.notice.Notifier
import io.wdsj.asw.bukkit.manage.permission.Permissions
import io.wdsj.asw.bukkit.manage.punish.Punishment
import io.wdsj.asw.bukkit.manage.punish.ViolationCounter
import io.wdsj.asw.bukkit.proxy.bungee.BungeeSender
import io.wdsj.asw.bukkit.proxy.velocity.VelocitySender
import io.wdsj.asw.bukkit.setting.PluginMessages
import io.wdsj.asw.bukkit.setting.PluginSettings
import io.wdsj.asw.bukkit.type.ModuleType
import io.wdsj.asw.bukkit.util.LoggingUtils
import io.wdsj.asw.bukkit.util.TimingUtils
import io.wdsj.asw.bukkit.util.Utils
import io.wdsj.asw.bukkit.util.cache.BookCache
import org.bukkit.ChatColor
import org.bukkit.entity.Player

class ASWBookPacketListener : PacketListenerAbstract(PacketListenerPriority.LOW) {
    override fun onPacketReceive(event: PacketReceiveEvent) {
        if (!AdvancedSensitiveWords.isInitialized) return
        if (!settingsManager.getProperty(PluginSettings.ENABLE_BOOK_EDIT_CHECK)) return
        val packetType = event.packetType
        val user = event.user
        val userName = user.name
        if (packetType === PacketType.Play.Client.EDIT_BOOK) {
            val player = event.getPlayer() as Player
            if (player.hasPermission(Permissions.BYPASS)) return
            var outMessage = ""
            var outList: List<String?> = ArrayList()
            val skipReturnLine = settingsManager.getProperty(PluginSettings.BOOK_IGNORE_NEWLINE)
            val isCacheEnabled = settingsManager.getProperty(PluginSettings.BOOK_CACHE)
            var shouldSendMessage = false
            val isCancelMode = settingsManager.getProperty(PluginSettings.BOOK_METHOD).equals("cancel", ignoreCase = true)
            val wrapper = WrapperPlayClientEditBook(event)
            val originalPages = wrapper.pages

            // Book content check
            val processedPages: MutableList<String> = ArrayList()
            val startTime = System.currentTimeMillis()
            for (i in 0 until originalPages.size) {
                var originalPage = originalPages[i]
                if (skipReturnLine) {
                    originalPage = originalPage.replace("\n", "").replace("§0", "")
                }
                if (settingsManager.getProperty(PluginSettings.PRE_PROCESS)) originalPage = originalPage.replace(Utils.getPreProcessRegex().toRegex(), "")
                val isBookCached = BookCache.isBookCached(originalPage)
                val censoredWordList = if (isBookCached && isCacheEnabled) BookCache.getCachedBookSensitiveWordList(originalPage) else AdvancedSensitiveWords.sensitiveWordBs.findAll(originalPage)
                if (censoredWordList.isNotEmpty()) {
                    val processedPage = if (isBookCached && isCacheEnabled) BookCache.getCachedProcessedBookContent(originalPage) else AdvancedSensitiveWords.sensitiveWordBs.replace(originalPage)
                    if (!isBookCached && isCacheEnabled) BookCache.addToBookCache(originalPage, processedPage, censoredWordList)
                    if (isCancelMode) {
                        event.isCancelled = true
                        shouldSendMessage = true
                        outMessage = originalPage
                        outList = censoredWordList
                        break
                    }

                    shouldSendMessage = true
                    outMessage = originalPage
                    outList = censoredWordList
                    processedPages.add(processedPage)
                } else {
                    processedPages.add(originalPages[i])
                }
            }
            if (!isCancelMode && shouldSendMessage) {
                wrapper.pages = processedPages
            }

            // Cross page check
            if (settingsManager.getProperty(PluginSettings.BOOK_CROSS_PAGE) && !shouldSendMessage) {
                var crossPageListString = originalPages.joinToString("").replace("\n", "").replace("§0", "")
                if (settingsManager.getProperty(PluginSettings.PRE_PROCESS)) {
                    crossPageListString = crossPageListString.replace(Utils.getPreProcessRegex().toRegex(), "")
                }
                val censoredWordListCrossPage = AdvancedSensitiveWords.sensitiveWordBs.findAll(crossPageListString)
                if (censoredWordListCrossPage.isNotEmpty()) {
                    shouldSendMessage = true
                    outMessage = crossPageListString
                    outList = censoredWordListCrossPage
                    event.isCancelled = true
                }
            }

            // Book title check
            var originalTitle = wrapper.title
            if (originalTitle != null) {
                if (settingsManager.getProperty(PluginSettings.PRE_PROCESS)) originalTitle = originalTitle.replace(Utils.getPreProcessRegex().toRegex(), "")
                val censoredWordListTitle = AdvancedSensitiveWords.sensitiveWordBs.findAll(originalTitle)
                if (censoredWordListTitle.isNotEmpty()) {
                    val processedTitle = AdvancedSensitiveWords.sensitiveWordBs.replace(originalTitle)
                    if (isCancelMode) {
                        event.isCancelled = true
                    } else {
                        wrapper.title = processedTitle
                    }
                    shouldSendMessage = true
                    outMessage = originalTitle
                    outList = censoredWordListTitle
                }
            }

            if (shouldSendMessage) {
                Utils.messagesFilteredNum.getAndIncrement()
                ViolationCounter.incrementViolationCount(player)
                if (settingsManager.getProperty(PluginSettings.HOOK_VELOCITY)) {
                    VelocitySender.sendNotifyMessage(player, ModuleType.BOOK, outMessage, outList)
                }
                if (settingsManager.getProperty(PluginSettings.HOOK_BUNGEECORD)) {
                    BungeeSender.sendNotifyMessage(player, ModuleType.BOOK, outMessage, outList)
                }
                val endTime = System.currentTimeMillis()
                TimingUtils.addProcessStatistic(endTime, startTime)
                if (settingsManager.getProperty(PluginSettings.NOTICE_OPERATOR)) {
                    Notifier.notice(player, ModuleType.BOOK, outMessage, outList)
                }
                if (settingsManager.getProperty(PluginSettings.BOOK_PUNISH)) {
                    AdvancedSensitiveWords.getScheduler().runTask {
                        Punishment.punish(player)
                    }
                }
                if (settingsManager.getProperty(PluginSettings.BOOK_SEND_MESSAGE)) {
                    player.sendMessage(ChatColor.translateAlternateColorCodes('&', AdvancedSensitiveWords.messagesManager.getProperty(PluginMessages.MESSAGE_ON_BOOK)))
                }
                if (settingsManager.getProperty(PluginSettings.LOG_VIOLATION)) {
                    LoggingUtils.logViolation(userName + "(IP: " + Utils.getPlayerIp(player) + ")(Book)", outMessage + outList)
                }
            }
        }
    }
}
